### PR TITLE
Refactor SESE loop canonicalization so that it is clear that undefs is inserted at preheader only

### PR DESCRIPTION
**Pull request by bgogul**
There is no change in functionality. This is just a code refactor to make subsequent code changes to undef handling is easy to implement and review.

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Review by mhong — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp [hunk 1]
```diff
@@ -502,6 +515,31 @@ SingleExitLoopTransformer::createNewLatch(SILBasicBlock *newHeader) {
   return latchBlock;
 }
 
+void SingleExitLoopTransformer::patchPreheader(SILBasicBlock* newHeader) {
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: Nit: `SILBasicBlock *`

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp [hunk 2]
```diff
@@ -316,9 +316,13 @@ class SingleExitLoopTransformer {
   /// Create a new latch block and clone all the arguments from new header.
   SILBasicBlock *createNewLatch(SILBasicBlock *newHeader);
 
-  // Patch the edges that go to the header and exit blocks to the new header or
-  // latch block as appropriate. Also, return the map consisting of indices
-  // assigned to the exit blocks.
+  /// Replace the preheader->header edge with preheader->newHeader edge
+  /// and update the arguments of preheader to match that of newheader.
+  void patchPreheader(SILBasicBlock* newHeader);
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: Nit: `SILBasicBlock *`

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp (hunk 1 above)
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: Done.

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp (hunk 2 above)
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: Done

**Pull request merged**